pkg/scanners/clairv4/convert.go

```diff
@@ -0,0 +1,187 @@
+// https://github.com/quay/clair/blob/v4.5.0/cmd/clairctl/manifest.go#L76.
+// It is assumed the given client handles auth.
+func fetchLayerURIAndHeader(client *http.Client, url, repository, digest string) (string, http.Header, error) {
+	path := fmt.Sprintf("/v2/%s/blobs/%s", repository, digest)
+	req, err := http.NewRequest(http.MethodGet, url+path, nil)
+	if err != nil {
+		return "", nil, errors.Wrap(err, "creating image layer pull request")
+	}
+	// We don't actually want the layer. We just want the headers.
+	req.Header.Add("Range", "bytes=0-0")
+	res, err := client.Do(req)
+	if err != nil {
+		return "", nil, errors.Wrap(err, "fetching image layer")
+	}
+	utils.IgnoreError(res.Body.Close)
+
+	res.Request.Header.Del("User-Agent")
+	res.Request.Header.Del("Range")
+
+	return res.Request.URL.String(), res.Request.Header, nil
+}
+
+// imageScan converts the given report to an image scan.
+func imageScan(report *claircore.VulnerabilityReport) *storage.ImageScan {
+	scan := &storage.ImageScan{
+		ScanTime:        gogotypes.TimestampNow(),
+		Components:      components(report),
+		OperatingSystem: os(report),
+	}
+
+	if scan.GetOperatingSystem() == "unknown" {
+		scan.Notes = append(scan.Notes, storage.ImageScan_OS_UNAVAILABLE)
+	}
+
+	return scan
+}
+
+func components(report *claircore.VulnerabilityReport) []*storage.EmbeddedImageScanComponent {
+	components := make([]*storage.EmbeddedImageScanComponent, 0, len(report.PackageVulnerabilities))
+	for id, pkg := range report.Packages {
+		vulnIDs := report.PackageVulnerabilities[id]
+		component := &storage.EmbeddedImageScanComponent{
+			Name:    pkg.Name,
+			Version: pkg.Version,
+			Vulns:   vulnerabilities(report.Vulnerabilities, vulnIDs),
+		}
+
+		components = append(components, component)
+	}
+
+	return components
+}
+
+// vulnerabilities returns a list of vulnerabilities from the given map indexed by the given ids.
+func vulnerabilities(vulnerabilities map[string]*claircore.Vulnerability, ids []string) []*storage.EmbeddedVulnerability {
+	// vulns will have at most len(ids) entries.
+	vulns := make([]*storage.EmbeddedVulnerability, 0, len(ids))
+	uniqueVulns := set.NewStringSet()
+	for _, id := range ids {
+		ccVuln := vulnerabilities[id]
+		if !uniqueVulns.Add(ccVuln.Name) {
+			// Already added this vulnerability, so ignore it.
+			continue
+		}
+
+		var publishedTime *gogotypes.Timestamp
+		if !ccVuln.Issued.IsZero() {
+			// Ignore the error, as publishedTime will just be `nil` if the given time is invalid.
+			publishedTime, _ = gogotypes.TimestampProto(ccVuln.Issued)
+		}
+		vuln := &storage.EmbeddedVulnerability{
+			Cve:               ccVuln.Name,
+			Summary:           ccVuln.Description,
+			Link:              link(ccVuln.Links),
+			PublishedOn:       publishedTime,
+			VulnerabilityType: storage.EmbeddedVulnerability_IMAGE_VULNERABILITY,
+			Severity:          normalizedSeverity(ccVuln.NormalizedSeverity),
+		}
```

Comment:
Where are the vectors added?

Reply:
Does the scanner enrich that data later in the scan workflow?

Reply:
I don't think Clair v4 gives CVSS score/vector @crozzy @hdonnay

Reply:
correct